### ISSUE TITLE
Simplify JobContext

### ISSUE THREAD

**Issue by onlyann** [procrastinate-org/procrastinate#1212]
The `JobContext` class mixes fields that are intended to be used by consumers of this library while other fields are used internally. 

As part of `v3`, it is a good time to clean up `JobContext` and only expose the fields that are intended for public use.
There is also the opportunity to simplify typing by removing the `task` field.

**Comment by medihack** [comment]
Closed by #1213

**Comment by ewjoachim** [comment]
(as seen in https://github.com/procrastinate-org/procrastinate/issues/1257#issuecomment-2551505079, maybe the task was useful after all...)

**Comment by onlyann** [comment]
> (as seen in https://github.com/procrastinate-org/procrastinate/issues/1257#issuecomment-2551505079, maybe the task was useful after all...)

Rather than reintroducing a circular reference with the `task`, how about adding useful fields directly on JobContext, such as the retry strategy instance in this case?

**Comment by ewjoachim** [comment]
I think it was pretty useful to have the task there in the end. Is there really no simple way we can make it work for typing ?

**Comment by onlyann** [comment]
Sorry, I don't have a better suggestion.

**Comment by medihack** [comment]
Then let's re-add the task to the JobContext and have a beta v3 release (it's the last issue left in v3 milestone). @onlyann, do you want to re-add it, or should I?

**Comment by onlyann** [comment]
Happy for you to add it @medihack.

Looking back at the codebase, the JobContext has the `task_name` and the `app` which is enough to look up the task without passing it in.

Effectively, I think you can simply add a new property/method that returns:

```
self.app.tasks.get(self.job.task_name)
```

**Comment by medihack** [comment]
Good point. Then, I don't see a need to do anything (maybe add a hint to the documentation on how to get the task) and keep the API simple. @ewjoachim Ok for you, too?

**Comment by ewjoachim** [comment]
Ok !

**Comment by medihack** [comment]
Cool. I will add a small hint about retrieving the task with its parameters to the documentation under the "Get more context for task execution" section and then close the issue.

**Comment by ewjoachim** [comment]
Closed as per #1283
